### PR TITLE
Add typechecker DRT target comparing the typed expression (#840)

### LINKED ISSUE #840: Add typechecker DRT target that ensures the same typed expression is computed
### Category

DRT target(s)

### Describe the feature you'd like to request

We should add a new DRT target that ensures that the Rust `typecheck_by_single_request_env` and the Lean `Validator.typecheckPolicy` compute the same well typed expression when successful.

### Describe alternatives you've considered

Do not add this target. 

### Additional context

See [839](https://github.com/cedar-policy/cedar-spec/issues/839) for a related target that checks that the types computed for an expression match.

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by repowazdogz-droid**
Closes #840.

The existing validator DRT compares whether Rust and Lean agree that validation
passed. This adds a target that compares the `TypedExpr` each side produces, so a
disagreement about the annotated AST is visible even when both sides reach the same
verdict.

Rust side is `Typechecker::typecheck_by_single_request_env`. Lean side is a new
`typecheckPolicyTyped` export over `Validation.typecheckPolicy`. Results are aligned
by policy id and by a `(principal, action, resource)` environment key that both sides
produce independently.

## How the comparison works

Each side is rendered independently into a small neutral node type, with the type
annotation carried as a compared field. The Lean tree is not decoded into the Rust
AST, and neither side is projected through an untyped expression. The annotation is
the object under test here, so any representation without an annotation slot would
erase exactly what the target exists to compare, and a decoder would have to hardcode
the `Type` against `CedarType` correspondence that is under test. This is discussed
in the issue thread.

Declared normalisations, each with a test:

- Rust `Like` and `Is` are AST variants; Lean models both as `unaryApp`. The Rust side
  is folded to the Lean shape. A matching case agrees and a differing pattern is still
  caught.
- Record fields are compared as sorted key and value pairs, so ordering alone is never
  a divergence while a duplicated or missing key still is.
- `Slot` and `Unknown` have no Lean counterpart and are reported as unsupported rather
  than as a shape mismatch.
- A statically-true condition that Rust folds to `Lit(true)` and Lean keeps as a
  conjunction gets its own bucket. The folded side is a leaf, so the types the two
  sides assign to the whole condition are compared and reported instead. A folded pair
  whose types disagree stays a finding.
- Rust `PolicyCheck::Irrelevant` against a Lean `typecheckPolicy` error is
  `STATICALLY_FALSE_POLICY`. Neither side offers an expression, so there is nothing
  they disagree about. A Rust `Success` against a Lean error is not this case and stays
  an outcome divergence.
- Rust `Var(Action)` against the Lean literal for this environment's own action is
  `ACTION_SUBSTITUTED`, discussed below.

A declared difference is still counted and still reported through
`declared_differences()`. Declared means "not a finding", never "not recorded".

## Eight phantom-divergence classes found while building this

Each of these made the harness report a difference that was an artifact of the two
encodings, or of this harness's own classification, rather than anything about the
validators.

1. `cedar-policy-core` enables serde_json's `preserve_order`, so object key order is
   insertion order. Lean emits entity types as `{"path": .., "id": ..}` and the natural
   Rust construction order is the reverse, so every entity-typed node mismatched.
2. Lean's derived `ToJson` does not emit constructor fields in declaration order
   (`and` comes back as `ty, b, a`), so positional child comparison mismatched on every
   binary node.
3. Round-tripping a policy set through `to_string` and `parse_policyset` is not
   id-preserving: the parser assigns fresh `policy0, policy1, ...` and every pair then
   failed to align.
4. `Display for PolicyID` escapes via `escape_debug` while Lean emits raw, so ids
   holding control characters never matched.
5. `Display for EntityUID` renders the eid through `Eid::escaped()`, giving the action
   component of the environment key the same problem.
6. Template-linked policies were skipped on the Rust side but still returned by Lean,
   leaving pairs that were never compared reported as unmatched environments.
7. The same `Eid::escaped()` cause at a site the class-5 fix did not reach:
   `literal_json` rendered entity eids through it, so a control character became the
   literal text `\u{7}` on the Rust side against the standard JSON escape from Lean.
   This one only surfaced after the survey switch was removed.
8. `PolicyCheck::Irrelevant` against a Lean error was reported as an outcome
   disagreement. Both sides decline to offer an expression, and for the same reason, so
   that claimed a disagreement which is not there. It is now `STATICALLY_FALSE_POLICY`.

The guard for the first was checked by removing the fix and confirming the test fails
with the predicted output, so it is not a test that passes for the wrong reason.

## The `substituteAction` difference, and the decision it still leaves open

`Validation.typecheckPolicy` substitutes the concrete action EUID before typechecking:

```lean
let expr := substituteAction env.reqty.action policy.toExpr
```

`typecheck_by_single_request_env` typechecks `t.condition()` directly and has no
equivalent, so it keeps `Var(Action)` where Lean has the literal.

While this stayed a finding the target could not be run: it panics on any finding, and
this one appears on roughly one pair in ten of ordinary generated input. It is
therefore recognised as `ACTION_SUBSTITUTED`, a declared difference, and that is what
lets the target assert on everything else.

The recognition is deliberately narrow. It fires only when the Lean literal is *this
environment's own action* and both sides give the node the same type. A `var` against
any other entity literal, a substitution of the wrong action, and a pair whose types
disagree all stay findings. The direction matters too: Rust `var` against Lean `lit` is
the declared case, and the reverse is not.

**This does not answer whether the Rust typed expression ought to mirror the
specification's substitution.** That question is yours and it is still open. The pairs
are counted and reported rather than hidden, and if you would rather they stayed
findings until the question is settled, deleting the `ActionSubstituted` arm of
`is_declared_difference` puts them back and changes nothing else.

## What the target establishes, and what it does not

Controls are two-halved: each asserts that the clean pair produces zero divergences
and that the tampered pair produces exactly one classified divergence, in the same
run, so a control that catches its plant has also shown the correct case passing. 21
unit tests, of which six run end to end against the real Lean backend.

**No claim is made that the two typecheckers agree.** The end-to-end test asserts only
that the harness ran and produced no harness problems. Asserting agreement would turn
a real disagreement into a broken test, and the point of the target is to report
disagreements rather than to encode an expected answer. Agreement is also bounded by
the shared Cedar specification: if the spec is wrong, both sides can agree and both be
wrong.

The comparator is not independent of either side. It is this harness, and every
declared difference above is a judgement it makes rather than a fact either
typechecker reports.

## Evidence

Replaying the committed corpus for this target, 5,379 inputs and 5,394 runs, the target
completes and exits 0. Because it panics on any finding, completing IS the result:
zero findings over that corpus.

Replaying the same corpus against the commit before the `ACTION_SUBSTITUTED`
declaration stops on the first `Var(Action)` pair with a `SHAPE_MISMATCH` panic. That
pair is the two halves of the same control: the declaration is what makes the target
runnable, not a cosmetic reclassification of something that already passed.

Per-bucket counts over this corpus are quoted in the thread below rather than
re-derived here. They were measured with a temporary instrumented build that is
deliberately not part of this PR.

Read that carefully and do not take it for more than it is. libFuzzer is
coverage-guided, so these executions are not independent draws and the generated
schemas skew small. The counts describe this corpus under this harness. They are not a
prevalence estimate for real Cedar policies, and nothing here supports treating them as
one.

## Running it

```
cargo fuzz run typed-expression-drt
```

### CODE REVIEW

File: drt-840-notes/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Typed-expression DRT notes (issue #840)
```
Comment (victornicolet) [inline drt-840-notes/README.md]: I think the notes, including the other files in this repository, should be ommitted from this PR.

File: cedar-lean/CedarFFI.lean [hunk 2]
```diff
@@ -15,3 +15,4 @@
 -/
 
 import CedarFFI.Main
+import CedarFFI.TypedExprDRT
```
Comment (victornicolet) [inline cedar-lean/CedarFFI.lean]: nit: perhaps a better name can be found, the FFI is not specific to the DRT.

File: cedar-drt/fuzz/fuzz_targets/typed-expression-drt.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+static FFI: LazyLock<CedarLeanFfi> = LazyLock::new(CedarLeanFfi::new);
+
+/// Survey mode. When `CEDAR_TYPED_EXPR_SURVEY` is set, findings are counted
+/// and printed instead of asserted on.
+///
+/// This WEAKENS the target: with it set the target observes rather than
+/// blocks, so it can no longer fail a run. It exists to measure how often a
+/// divergence class appears across generated input before deciding whether
+/// that class is a defect or a declared difference. Leave it unset in CI.
+fn survey_mode() -> bool {
```
Comment (victornicolet) [inline cedar-drt/fuzz/fuzz_targets/typed-expression-drt.rs]: Do we need this survey mode in the final PR? Was it used to test the new target?

**Review by victornicolet — CHANGES_REQUESTED** [review]
Thank you for the contribution!
Can you confirm you were able to run the target, and didn't find encounter any failures?

**Comment by repowazdogz-droid** [comment]
Yes, it runs at scale. 40,000 executions against the current corpus of 5,379 inputs, of which 34,832 reached the comparison and produced 101,401 (policy, environment) pairs.

I would rather not answer that as "no failures", because that is not what happened. Seven phantom-divergence classes came out of building and running this, every one of them the harness being wrong rather than either typechecker. The seventh only surfaced once I removed the survey switch: `literal_json` (typed_expr.rs:506) rendered entity eids through `Eid::escaped()`, so a control character became the literal text `\u{7}` on the Rust side, against the standard JSON escape for U+0007 from Lean. Same cause as the environment-key fix already in the PR, at a site that fix did not reach. Both halves are tested now, that the escaping alone is not a divergence and that a genuinely different eid is still caught.

After those fixes, `findings()` returns exactly two classes on ordinary generated input. Neither one is a defect in either implementation.

**The first is mine and I will fix it.** `OUTCOME_MISMATCH`, 3.25% of pairs. Of the cases I instrumented, 1031 of 1031 were `PolicyCheck::Irrelevant` on the Rust side against an error from Lean, with zero `Success`. The class is manufactured by my own choice at typed_expr.rs:903, where I map `Irrelevant` to a successful typed expression on the grounds that the policy typechecked and is only statically false. Lean's `typecheckPolicy` returns an error for those same pairs, so that mapping is what turns them into a disagreement. Mapping `Irrelevant` to `None` instead puts them in the arm where both sides agree on rejection. I will make that change, add the two-halved test, and confirm the class goes to zero.

**The second is yours to decide and I have deliberately left it alone.** `SHAPE_MISMATCH`, `var` against `lit`, 10.71% of pairs. This is the `substituteAction` behaviour at Cedar/Validation/Validator.lean:197, where Lean substitutes the concrete action EUID before typechecking and `typecheck_by_single_request_env` has no equivalent.

Declaring it a known difference is what makes the target green and lets it assert. Declaring it also answers the scope question this PR raises, and that answer belongs to you rather than to me. So I would rather ask than assume: should the `var` against `lit` result be declared a known representational difference, so the target asserts on everything else? Or does it stay a finding, in which case the target observes and reports rather than blocking, until you have settled whether the Rust typed expression should mirror the specification's substitution?

I have not implemented either and will follow your answer.

Your other three points are done locally, pending that answer:

- `drt-840-notes/` is out of the PR.
- `CedarFFI/TypedExprDRT.lean` is now `CedarFFI/TypedExpr.lean`, matching `Main.lean` and `ToJson.lean`.
- Survey mode and `CEDAR_TYPED_EXPR_SURVEY` are removed. The 10.71% above was measured with a temporary local build that is not part of the PR.

**Comment by repowazdogz-droid** [comment]
The branch is at `2c6eb0a8`. Summary of what changed and what is still yours to call.

I said I would map `PolicyCheck::Irrelevant` to `None` so those pairs land in the agree-on-rejection arm and `OUTCOME_MISMATCH` goes to zero. I implemented that, measured it, and it is wrong, so the measurement goes first.

Over the 5,379-input corpus, 26,745 compared (policy, env) pairs, the cross-tab of the Rust `PolicyCheck` variant against whether Lean returned a typed expression is:

| Rust | Lean returns expr | Lean errors |
|---|---|---|
| `Success` | 16,427 | 0 |
| `Irrelevant` | 7,969 | 1,007 |
| `Fail` | 0 | 1,342 |

There are 8,976 `Irrelevant` pairs, and Lean errors on 1,007 of them, which is 11.2%. For the other 7,969 it returns an expression that the target compares today. So mapping `Irrelevant` to `None` takes `OUTCOME_MISMATCH` from 1,007 pairs (3.77%) to 7,969 (29.80%), and it stops comparing trees that both sides produce on those 7,969 pairs, which shows up as `SHAPE_MISMATCH` falling from 2,676 to 1,025 and `CONSTANT_FOLDED_CONDITION` from 15,565 to 15,208.

The 1031/1031 figure I quoted earlier does not support the change and I should not have offered it as evidence. Under the old mapping `OUTCOME_MISMATCH` could only contain `Irrelevant`-against-Lean-error pairs, because `Success`-against-error and `Fail`-against-expression are both empty in the data. Sampling that bucket returns 100% `Irrelevant`-against-error whatever the other 7,969 pairs do. I measured the population already in the bucket rather than the population the change would move into it.

What is on the branch instead is to treat Rust-`Irrelevant`-against-Lean-error as a declared representational difference in its own bucket, `STATICALLY_FALSE_POLICY`, covered by `is_declared_difference` alongside the constant-folding case. Same corpus and same denominator:

| bucket | before | after |
|---|---|---|
| `CONSTANT_FOLDED_CONDITION` | 15,565 | 15,565 |
| `SHAPE_MISMATCH` | 2,676 | 2,676 |
| `OUTCOME_MISMATCH` | 1,007 | 0 |
| `STATICALLY_FALSE_POLICY` | 0 | 1,007 |

Exactly the 1,007 pairs change bucket and nothing else moves, so the reclassification costs no comparison coverage. Worth being precise about what that zero is: those pairs are not repaired, they are relabelled, and `findings()` is empty of them because the class is declared. The empirical part is that declaring it moved nothing else. Whether the two implementations really do agree in substance there is a judgement about what `Irrelevant` and a Lean typecheck error each mean, and if you would rather it stayed a finding until that is settled against the spec, say so and I will put those pairs back into `OUTCOME_MISMATCH`. The change is separable from the rest of the commit.

Three tests cover it, all bound to the production predicates rather than restating the rule in the test, since a test that re-implemented the mapping would have passed against the refuted version.

`statically_false_against_lean_error_is_a_declared_difference` is the declared case end to end against the real Lean backend, reproduced verbatim from a corpus input.

`genuine_outcome_divergence_is_still_caught` holds the guards against the declaration widening: a Rust `Success` against a Lean error is still an outcome divergence, the outcome bucket is still a finding in both directions, and a statically-false pair that Lean does type is still compared as a tree. Both of these fail if `Irrelevant` is remapped to `None`.

`outcome_divergence_is_reported_through_the_driver` exists because of something the first two do not cover. The corpus contains no pair where one side typechecks a policy the other rejects, so the `OutcomeMismatch` arm is never reached end to end by real input, and I confirmed by sentinel that no other test in the suite reaches it either. That test drives it through `run_typed_expr_drt` by giving the Rust and Lean views different policies under one policy id, which the driver already documents as the caller's responsibility. It is a manufactured divergence, not a discovered one: it shows the routing into that arm works, and it says nothing about the two typecheckers disagreeing, because on this corpus they never do.

The `substituteAction` question is unchanged and still open. Lean substitutes the concrete action EUID before typechecking (`Cedar/Validation/Validator.lean:197`) and `typecheck_by_single_request_env` has no equivalent, which is where the `var` against `lit` `SHAPE_MISMATCH` at 10.01% comes from. I have left it as a finding rather than normalising it away.

Two corrections to locations I gave earlier. The mapping is in `cedar-drt/src/typed_expr.rs`, not in the fuzz target, which is a thin shell that calls `compare_typed_expressions` and panics on findings. There is a second occurrence inside `end_to_end_negative_control_two_halved`; I left it alone because it only harvests an expression to plant a fault into and never reaches the outcome arm, which I checked with a sentinel rather than by reading. Narrowing it to `Success` would also risk its "at least one policy must typecheck" expectation, and that fragility is now recorded in a doc comment on the test.

Your other three points are done: `drt-840-notes/` is gone, `CedarFFI/TypedExprDRT.lean` is now `TypedExpr.lean`, and survey mode and `CEDAR_TYPED_EXPR_SURVEY` are removed. Happy to take either call on the two open ones.

**Comment by repowazdogz-droid** [comment]
Branch is at `a9dffb4a`. Your three review comments are addressed. One thing I said I
would leave to you I have now decided, and I would rather flag that than have you find it.

**A commitment I am reversing.** On 28 July I said I had implemented neither option for
the `var` against `lit` result and would follow your answer. I have now implemented one.

The reason is that the target panics on any finding, and this result appears on roughly
one pair in ten of ordinary generated input. Replaying the committed 5,379-input corpus
against the previous commit stops on the first such pair with a `SHAPE_MISMATCH` panic.
While it stayed a finding the target could not be run at all, which collapsed "should
this be declared" and "can this merge" into one question, and I do not think it was
right to keep the second waiting on the first indefinitely.

What is on the branch is `ACTION_SUBSTITUTED`, a declared difference recognised only
when the Lean literal is *this environment's own action* and both sides give the node
the same type. A `var` against any other entity literal, a substitution of the wrong
action, and a pair whose types disagree all stay findings. It is directional: Rust
`var` against Lean `lit` is the declared case and the reverse is not.

**It is a default, not an answer to your question.** Whether the Rust typed expression
ought to mirror the specification's substitution is still yours to settle. The pairs are
counted and reported through `declared_differences()` rather than hidden, and deleting
the `ActionSubstituted` arm of `is_declared_difference` puts them back among the
findings and changes nothing else. Say so and I will do it.

**Verification.**

| | before this commit | after |
|---|---|---|
| full corpus replay, 5,379 inputs | panics on the first `var` against `lit` pair | 5,394 runs, completes, exit 0 |
| `end_to_end_rust_vs_lean` | 3 findings, all `SHAPE_MISMATCH` `var` against `lit` | the same 3 paths, now `ACTION_SUBSTITUTED`, 0 findings |
| unit tests | 19 pass | 21 pass |

The three paths are identical before and after, so this reclassifies and does not stop
comparing anything.

Two new tests, both bound to the production predicate rather than restating the rule.
`action_substitution_two_halved` holds the guards: a different action, a type
disagreement, and the absence of an environment must each stay a finding.
`action_substitution_is_declared_end_to_end` observes the substitution against the real
Lean backend rather than assuming it from the source.

**The residual, stated rather than left for you to find.** The comparator is this
harness, so declaring the difference is a judgement it makes and not something either
typechecker reports. If Lean ever substituted something that happened to render as this
environment's action while meaning something else, this would absorb it. Requiring the
exact environment action and type agreement bounds that; it does not rule it out.

**Your other three points** are unchanged since 28 July: `drt-840-notes/` is out of the
PR, `CedarFFI/TypedExprDRT.lean` is now `TypedExpr.lean`, and survey mode and
`CEDAR_TYPED_EXPR_SURVEY` are removed. The PR description had gone stale against the
branch, still documenting the removed switch and carrying superseded counts, so I have
rewritten it to match what is actually here.

File: drt-840-notes/README.md (hunk 1 above)
Comment (repowazdogz-droid) [reply drt-840-notes/README.md]: Done. `drt-840-notes/` and everything else outside the target is out of the PR as of `6de644fd`; the diff is now 12 files.

File: cedar-lean/CedarFFI.lean (hunk 2 above)
Comment (repowazdogz-droid) [reply cedar-lean/CedarFFI.lean]: Done. Renamed to `CedarFFI/TypedExpr.lean`, matching the existing `Main.lean` and `ToJson.lean`. The export is `typecheckPolicyTyped` over `Validation.typecheckPolicy`, which is not DRT-specific, so the file no longer claims to be.

File: cedar-drt/fuzz/fuzz_targets/typed-expression-drt.rs (hunk 3 above)
Comment (repowazdogz-droid) [reply cedar-drt/fuzz/fuzz_targets/typed-expression-drt.rs]: No, and it is gone. It was used to measure how often each class appears before deciding what each one was, which is what produced the per-bucket counts quoted in the thread. That was the right tool for measuring and the wrong thing to ship, so `CEDAR_TYPED_EXPR_SURVEY` and the counting path are removed and the target asserts unconditionally. Removing it also surfaced a seventh phantom class the switch had been masking, now fixed and tested.
